doc/api/async_hooks.md

```diff
@@ -682,6 +682,172 @@ never be called.
+  } else {
+    console.log(...args);
+  }
+}
+
+http.createServer((request, response) => {
+  asyncContext.run((store) => {
+    store.set(kReq, request);
+    someAsyncOperation((err, result) => {
+      if (err) {
+        log('ERR', err.message);
+      }
+    });
+  });
+})
+.listen(8080);
+```
+
+### Choosing between asynchronous and synchronous methods
+
+AsyncContext offer three pair of methods to enter and exit a context:
+* `run/exit`
+* `runAndReturn/exitAndReturn`
+* `enterSync/exitSync`
+
+#### When to choose `run/exit`
+These methods are asynchronous. They are called with a callback function that
+runs within a new asynchronous call. This is the most explicit behavior as
+everything that is executed within the callback of `run` (including further
+asynchronous operations) will have access to the store. The `exit` method has
+a similar behavior to stop context propagation.
+
+If an instance of AsyncContext is used for error management (for instance with
+`process.setUncaughtExceptionCaptureCallback`). Only the exceptions thrown
+in the scope of the callback function will be associated with the context.
+
+These methods are the safest to call as they provide strong scoping.
+
+#### When to choose `runAndReturn/exitAndReturn`
+These methods are synchronous and scoped. The callback function will be executed
+synchronously and it's return value will be returned by `runAndReturn`
+and `exitAndReturn`. The store will only be accessible from within the callback
+function and the asynchronous operations created within this scope.
+If the callback throws an error, `runAndReturn` and `exitAndReturn` will throw
+it too.
+
+These methods provide good scoping while being synchronous.
+
+#### What are the risks of `enterSync/exitSync`
+These methods are to be used only if you know what you are doing. After calling
+`enterSync`, the context will be propagated to everything that follows.
+For instance, in the case of an event emitter, if the first executed listener
+calls `enterSync`, the following ones will have access to the store.
+```js
+emitter.on('event', () => {
+    asyncContext.enterSync();
+});
+emitter.on('event', () => {
+    asyncContext.getStore(); // will return  the store.
+});
+```
+If an instance of AsyncContext is used for error management (for instance with
+`process.setUncaughtExceptionCaptureCallback`). All exceptions will be
+associated with the context.
+
+These methods are less safe than `run/exit` and `runAndReturn/exitAndReturn` as
+they have no scope.
+
+### new AsyncContext()
+
+* Returns: {Object}
+
+Creates a new instance of `AsyncContext`. Storage is only provided inside a `run`
+ method call.
+
+### asyncContext.run(callback)
+
+* `callback` {Function}
+  * `store` {Map}
+
+Calling `asyncContext.enter(callback)` will create a new asynchronous context.
+This method will call the callback with a `Map` known as the store as argument.
+The callback will be ran asynchronously.
```

Comment:
Maybe add some comment that any exception thrown by `callback` is unhandled.

Reply:
Good point, will do